### ISSUE TITLE
no-unused-expression gives lint error on expressions like `o && o.foo()`

### ISSUE THREAD

**Issue by gmoothart** [palantir/tslint#1413]
### Bug Report
- **TSLint version**: `3.13.0`
- **TypeScript version**: `1.8.10`
- **Running TSLint via**: CLI
#### TypeScript code being linted

``` ts
  let el: ng.IAugmentedJQuery;
  ...
  afterEach(() => {
    el && el.remove();
  });
```

with `tslint.json` configuration:

``` json
{
    "rules": {
        "no-unused-expression": true
    }
}
```
#### Actual behavior

test.ts[67, 5]: expected an assignment or function call - on the line `el && el.remove();`
#### Expected behavior

I expected this to not produce any errors. It seems like a valid usage and perfectly readable.

**Comment by IllusionMH** [comment]
Something like `a && <call expression>` or even `a && b && ... && <call expression>` are easy to fix.

@adidahiya, @JKillian is there any need for support of `a || <call expression>` or different combinations with parenthesis something like `(a || required('a')) && (b || required('b'))`?

**Comment by JKillian** [comment]
It's a little tricky, because `someObj.someFunc()` may not have side-effects and the expression may actually be unused. However, I'd okay with a feature that allows the use cases you mentioned above @IllusionMH.

(Sidenote: what's the correct term for a function that is without side-effects but not necessarily pure?)

**Comment by IllusionMH** [comment]
Fixed with #1638 - `allow-fast-null-checks` option added.
